proposals/mpi-operator-proposal.md

```diff
@@ -0,0 +1,224 @@
+spec:
+  gpus: 128
+  template:
+    spec:
+      containers:
+      - name: tensorflow-benchmarks
+        image: rongou/tensorflow_benchmarks:latest
+```
+
+For more flexibility, user can choose to specify the resources explicitly:
+```yaml
+apiVersion: kubeflow.org/v1alpha1
+kind: MPIJob
+metadata:
+  name: tensorflow-benchmarks
+spec:
+  replicas: 16
+  template:
+    spec:
+      containers:
+      - name: tensorflow-benchmarks
+        image: rongou/tensorflow_benchmarks:latest
+        resources:
+          limits:
+            nvidia.com/gpu: 8
+```
+
+Either case would result in a worker `StatefulSet` and a launcher `Job`.
+
+### Resulting Worker
+```yaml
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: ${job-id}-worker
+spec:
+  podManagementPolicy: Parallel
+  replicas: 15
+  selector:
+    matchLabels:
+      app: ${job-id}-worker
+  serviceName: ${job-id}-worker
+  template:
+    metadata:
+      labels:
+        app: ${job-id}-worker
+    spec:
+      containers:
+      - name: tensorflow-benchmarks
+        image: rongou/tensorflow_benchmarks:latest
+        command: ["sleep"]
+        args: ["365d"]
+        resources:
+          limits:
+            nvidia.com/gpu: 8
+        volumeMounts:
+        - name: ml-job-config
+          mountPath: /etc/mpi
+      volumes:
+      - name: ml-job-config
+        configMap:
+          name: ${job-id}-config
+          items:
+          - key: kubexec.sh
+            path: kubexec.sh
+            mode: 365
+```
+
+### Resulting Launcher
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: ${job-id}-launcher
+spec:
+  template:
+    spec:
+      initContainers:
+      - name: kubectl-delivery
+        image: rongou/kubectl-delivery:latest
+        volumeMounts:
+        - name: ml-job-kubectl
+          mountPath: /opt/kube
+        env:
+        - name: TARGET_DIR
+          value: /opt/kube
+      containers:
+      - name: tensorflow-benchmarks
+        image: rongou/tensorflow_benchmarks:latest
+        resources:
+          limits:
+            nvidia.com/gpu: 8
+        volumeMounts:
+        - name: ml-job-kubectl
+          mountPath: /opt/kube
+        - name: ml-job-config
+          mountPath: /etc/mpi
+        env:
+        - name: OMPI_MCA_plm_rsh_agent
+          value: /etc/mpi/kubexec.sh
+        - name: OMPI_MCA_orte_default_hostfile
+          value: /etc/mpi/hostfile
+      restartPolicy: OnFailure
+      serviceAccountName: ${job-id}-launcher
+      volumes:
+      - name: ml-job-kubectl
+        emptyDir: {}
+      - name: ml-job-config
+        configMap:
+          name: ${job-id}-config
+          items:
+          - key: kubexec.sh
+            path: kubexec.sh
+            mode: 365
+          - key: hostfile
+            path: hostfile
+            mode: 292
+```
+
+The launcher pod invokes `mpirun` and communicates with worker pods through MPI.
+The initial handshake is done through `kubectl exec` instead of SSH. Logs can
```

Comment:
SSH-less(`kubectl exec`-based) mpi is fantastic 👍 !! I really love it!!

- It perhaps similar to @alsrgv pointed out, but, it would have to detect proper kubernetes version from the conatiner(side-car container?) because kubernetes cluster woud be upgraded in time. 
- Would kubernetes api server be overloaded if large openmpi jobs are spawned and communicated heavily each other??

Reply:
Yes，I think the pain points is ssh,if can SSH-less it will very fantastic

Reply:
The `kubectl` binary is really only used for the `pods/exec` api call, so version skew might not be too big of an issue. We could write a standalone go binary, but it'd be replicating the `kubectl` functionality, so I'm not sure if it's worth it. If someone can figure out how to call `pods/exec` from a shell script, maybe we can switch to that. :)